File: Div fra boka/15.1-cut-rod.jl


function cutrod(p,n)
	memo = Dict(0 => 0)
	function helper(n)
		if haskey(memo,n)
			return memo[n]
		end
		q = -1
		for i in 1:n
			q = max(q,p[i]+helper(n-i))
		end
		memo[n] = q
		return q
	end
	return helper(n)

end


n = 4
p = [4,5,12,15]
if n > length(p)
		append!(p,zeros(Int,n-length(p)))
end
print(cutrod(p,n))

File: Div fra boka/15.1-cut-rod.py
def cutrod(p, n):
	if n <= 0:
		return 0
	q = -float('inf')
	for i in range(0, n):
		q = max(q, p[i] + cutrod(p, n-i-1))
	
	return q

def memoized_c

n = 20
p = [1, 5, 8, 9, 10, 17, 17, 20, 24, 30]
if len(p) <= n:
	p.extend([0] * (n - len(p)))

